### PR TITLE
MAINT: Start applying ruff/flake8-comprehensions rules (C4)

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
In preparation for https://github.com/numpy/numpy/issues/24994.

### CODE REVIEW

**Comment by mhvk** [comment]
This also has linting errors! (So I didn't review yet)

**Comment by DimitriPapadopoulos** [comment]
The linting errors are not caused by existing code, not by my changes. I would like to add `noqa: E231` in this specific case.  Thoughts?

**Comment by mhvk** [comment]
Probably best to just fix them - best not to merge PRs which do not pass the tests...

**Comment by DimitriPapadopoulos** [comment]
Should be fixed by running ` ruff check --select E231` on this specific file (requires `--preview` by the way): 
```console
$ ruff check --preview --select E231 numpy/_core/tests/test_nditer.py --fix
Found 33 errors (33 fixed, 0 remaining).
$ 
```

File: numpy/_core/tests/test_cython.py [hunk 1]
```diff
@@ -217,10 +217,8 @@ def test_multiiter_fields(install_temp, arrays):
     assert bcast.shape == checks.get_multiiter_shape(bcast)
     assert bcast.index == checks.get_multiiter_current_index(bcast)
     assert all(
-        [
-            x.base is y.base
+        x.base is y.base
             for x, y in zip(bcast.iters, checks.get_multiiter_iters(bcast))
```
Comment (mhvk) [inline numpy/_core/tests/test_cython.py:221]: Am surprised that this weird indentation passes the linter...

File: numpy/_core/tests/test_cython.py [hunk 2]
```diff
@@ -280,10 +278,8 @@ def test_npyiter_api(install_temp):
         x is y for x, y in zip(checks.get_npyiter_operands(it), it.operands)
     )
     assert all(
-        [
-            np.allclose(x, y)
+        np.allclose(x, y)
             for x, y in zip(checks.get_npyiter_itviews(it), it.itviews)
```
Comment (mhvk) [inline numpy/_core/tests/test_cython.py:282]: And another one.

**Review by mhvk — COMMENTED** [review]
Hmm, weird that that was the problem, especially as there seem to be worse problems (at least visually).

File: numpy/ma/tests/test_core.py [hunk 3]
```diff
@@ -3789,19 +3789,19 @@ def test_toflex(self):
         assert_equal(record['_mask'], data._mask)
 
         ndtype = [('i', int), ('s', '|S3'), ('f', float)]
-        data = array([(i, s, f) for (i, s, f) in zip(np.arange(10),
+        data = array(list(zip(np.arange(10),
                                                      'ABCDEFGHIJKLM',
-                                                     np.random.rand(10))],
+                                                     np.random.rand(10))),
```
Comment (charris) [inline numpy/ma/tests/test_core.py:3794]: This could use some cleanup.

File: numpy/ma/tests/test_core.py [hunk 4]
```diff
@@ -3789,19 +3789,19 @@ def test_toflex(self):
...
         record = data.toflex()
         assert_equal(record['_data'], data._data)
         assert_equal(record['_mask'], data._mask)
 
         ndtype = np.dtype("int, (2,3)float, float")
-        data = array([(i, f, ff) for (i, f, ff) in zip(np.arange(10),
+        data = array(list(zip(np.arange(10),
                                                        np.random.rand(10),
-                                                       np.random.rand(10))],
+                                                       np.random.rand(10))),
```
Comment (charris) [inline numpy/ma/tests/test_core.py:3804]: This could use some cleanup.

**Review by charris — CHANGES_REQUESTED** [review]
Couple of ugly spots, otherwise looks OK. A few tricky bits, but mostly straightforward.

File: numpy/ma/tests/test_core.py (hunk 4 above)
Comment (DimitriPapadopoulos) [reply numpy/ma/tests/test_core.py:3804]: You mean indentation cleanup, don't you?

**Review by charris — APPROVED** [review]

**Comment by charris** [comment]
Thanks @DimitriPapadopoulos .

**Pull request merged**